### ISSUE TITLE
Nebula broken

### ISSUE THREAD

**Issue by stinerman** [yt-dlp/yt-dlp#3848]
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.05.18** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

United States

### Description

It seems that Nebula support is broken. This seems to be a problem with all videos. I know it worked last week because I verified that the change in [Issue 3609](https://github.com/yt-dlp/yt-dlp/issues/3609) worked for me. I'm not on a particular release but did make sure I did a pull from master before I submitted the report.

I've verified that I can watch the videos directly via the website with no problems.

### Verbose log

```shell
stinerman@pc-debian:~/Videos$ ~/bin/yt-dlp -vU https://nebula.app/videos/wendover-how-to-design-a-theme-park-to-take-tons-of-your-money
[debug] Command-line config: ['-vU', 'https://nebula.app/videos/wendover-how-to-design-a-theme-park-to-take-tons-of-your-money']
[debug] User config "/home/stinerman/.config/yt-dlp/config": ['--mark-watched', '--netrc', '--prefer-insecure', '--no-check-certificate']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.05.18 [b14d52355] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: d9473db78
[debug] Python version 3.9.2 (CPython 64bit) - Linux-5.10.0-14-amd64-x86_64-with-glibc2.31
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 4.3.4-0, ffprobe 4.3.4-0
[debug] Optional libraries: Cryptodome-3.9.7, certifi-2020.06.20, secretstorage-3.3.1, sqlite3-2.6.0
[debug] Proxy map: {}
Latest version: 2022.05.18, Current version: 2022.05.18
yt-dlp is up to date (2022.05.18)
[Nebula] Logging in to Nebula with supplied credentials
[Nebula] Authorizing to Nebula
[Nebula] Retrieving Zype access token
[debug] [Nebula] Extracting URL: https://nebula.app/videos/wendover-how-to-design-a-theme-park-to-take-tons-of-your-money
[Nebula] wendover-how-to-design-a-theme-park-to-take-tons-of-your-money: Fetching video meta data
ERROR: wendover-how-to-design-a-theme-park-to-take-tons-of-your-money: An extractor error has occurred. (caused by KeyError('zype_id')); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "/home/stinerman/src/yt-dlp/yt_dlp/extractor/common.py", line 642, in extract
    ie_result = self._real_extract(url)
  File "/home/stinerman/src/yt-dlp/yt_dlp/extractor/nebula.py", line 247, in _real_extract
    return self._build_video_info(video)
  File "/home/stinerman/src/yt-dlp/yt_dlp/extractor/nebula.py", line 119, in _build_video_info
    zype_id = episode['zype_id']
KeyError: 'zype_id'
```

**Comment by stinerman** [comment]
FYI on this @hheimbuerger. This is a friendly ping because I know you do a lot of the maintenance. Thank you for that.

**Comment by hheimbuerger** [comment]
@stinerman Thank you for tagging me, I'll take a look asap!

**Comment by hheimbuerger** [comment]
@stinerman I can reproduce this. Even the unit tests all fail now, so it's not just a mistake I made, but rather Nebula changed something in their architecture.

I did a quick analysis. The good news is that the channel support seems fully functional (it just can't extract any videos, but for the same reason direct video downloads do not work anymore), and the metadata extraction appears completely intact, too.

But it seems that Nebula no longer directly exposes the Zype video ID, which the Nebula extractor currently uses to hand video downloading off to the Zype extractor.
Instead, there's now a `/video/{video_id}/stream/` endpoint, which returns a manifest URL. The manifest URL returns an M3U file. So probably, this is the point where we can drop the Zype extractor dependency completely and instead directly hand off the manifest for downloading.

It all looks very clean and well-structured on the Nebula API, so I should be able to do this in the next days. But yeah, it does require a code change to restore Nebula support.

tl;dr: Nebula video extractor currently fully broken. Fix is possible, PR will come in the next days.

**Comment by hheimbuerger** [comment]
I have a roughly working branch on https://github.com/hheimbuerger/yt-dlp/tree/fix-nebula-extractor

Please note that because previously, Nebula had no video IDs and we were therefore using the Zype video IDs, and now the Zype video ID is no longer clearly exposed (I found it somewhere as one component in a returned URL, together with some other IDs I cannot identify — but I don't think it's appropriate to parse it out of that), all Nebula videos now have different IDs, as far as yt-dlp is concerned.
This will 'break' archive file support, as previously downloaded videos will now get downloaded again.

For this reason and other insecurities, I cannot really consider this PR-ready yet. Unfortunately, I'll be mostly unavailable for the next ten days, and won't really be able to work on this. 

But if anyone wants to test this and send feedback in the meantime, or review the code and give feedback on that (I somewhat randomly found the `_extract_m3u8_formats_and_subtitles` helper function, no idea whether this is the best approach to parse an M3U…), that would be highly appreciated.

**Comment by pukkandan** [comment]
It is generally preferred not to break archives. But if the Zype ID is likely to disappear from the available data altogether in the near future, we'll need to break archive sooner or later. In that case, might as well do it now

**Comment by eladkarako** [comment]
are you also providing account details here? (you haven't specified it).  

normally you can either provide credentials for your account  
or clear all cookies and and reset your dynamic ISP connection to get a new IP  
which will then put you in a page named "first one is on us" without login.  

`ERROR: [Nebula] wendover-how-to-design-a-theme-park-to-take-tons-of-your-money: This video is only available for registered users. Use --cookies, --cookies-from-browser, --username and --password, or --netrc (watchnebula) to provide account credentials`

<details><summary>log</summary>

```txt

D:\DOS\yt-dlp_kit>call "D:\DOS\YT-DLP~1\resources\WPy32\python-3.10.2\python.exe" "D:\DOS\YT-DLP~1\resources\WPy32\python-3.10.2\Lib\site-packages\yt_dlp" "https://nebula.app/videos/wendover-how-to-design-a-theme-park-to-take-tons-of-your-money" --ignore-config --config-location "D:\DOS\YT-DLP~1\yt-dlp_conf_base.conf" --merge-output-format mkv --remux-video mkv 
[debug] Command-line config: ['https://nebula.app/videos/wendover-how-to-design-a-theme-park-to-take-tons-of-your-money', '--ignore-config', '--config-location', 'D:\\DOS\\YT-DLP~1\\yt-dlp_conf_base.conf', '--merge-output-format', 'mkv', '--remux-video', 'mkv']
[debug] | Config "D:\DOS\YT-DLP~1\yt-dlp_conf_base.conf": ['--force-ipv4', '--windows-filenames', '--sleep-requests', '1', '--sleep-interval', '1', '--sleep-subtitles', '2', '--max-sleep-interval', '3', '--console-title', '--no-check-certificates', '--prefer-insecure', '--user-agent', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4599.0 Safari/537.36', '--concurrent-fragments', '3', '--skip-unavailable-fragments', '--embed-subs', '--write-subs', '--write-auto-subs', '--sub-langs', 'en,iw,iw-en,ja,ko', '--embed-metadata', '--embed-chapters', '--embed-thumbnail', '--convert-thumbnails', 'jpg', '--sponsorblock-mark', 'intro,outro,poi_highlight', '--sponsorblock-remove', 'sponsor,selfpromo,interaction,filler,music_offtopic', '--sponsorblock-api', 'https://sponsor.ajay.app', '--no-abort-on-error', '--skip-unavailable-fragments', '--verbose', '--print-traffic']
[debug] Encodings: locale cp1255, fs utf-8, pref cp1255, out cp1255 (No ANSI), error cp1255 (No ANSI), screen cp1255 (No ANSI)
[debug] yt-dlp version 2022.05.18 [b14d52355]
[debug] Python version 3.10.2 (CPython 32bit) - Windows-10-10.0.19044-SP0
[debug] Checking exe version: ffprobe -bsfs
[debug] Checking exe version: ffmpeg -bsfs
[debug] exe versions: ffmpeg N-106943-g41a558fea0-2022-05-19-nonfree (fdk,setts), ffprobe N-106943-g41a558fea0-2022-05-19-nonfree, phantomjs 2.5.0, rtmpdump 2.3
[debug] Optional libraries: Cryptodome-3.14.1, brotli-1.0.9, certifi-2021.10.08, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.2
[debug] Proxy map: {}
[debug] [Nebula] Extracting URL: https://nebula.app/videos/wendover-how-to-design-a-theme-park-to-take-tons-of-your-money
[Nebula] wendover-how-to-design-a-theme-park-to-take-tons-of-your-money: Fetching video meta data
[Nebula] Reauthenticating to Nebula and retrying, because last bearer call resulted in error 403
send: b'GET /video/wendover-how-to-design-a-theme-park-to-take-tons-of-your-money/ HTTP/1.1\r\nHost: content.watchnebula.com\r\nAuthorization: Bearer None\r\nUser-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4599.0 Safari/537.36\r\nAccept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8\r\nAccept-Language: en-us,en;q=0.5\r\nSec-Fetch-Mode: navigate\r\nAccept-Encoding: gzip, deflate, br\r\nConnection: close\r\n\r\n'
reply: 'HTTP/1.1 403 Forbidden\r\n'
header: Allow: GET, HEAD, OPTIONS
header: Content-Type: application/json
header: Date: Fri, 27 May 2022 08:56:21 GMT
header: ETag: "442d8d4f8a2f183a0cb2dbbe28f084ce"
header: Referrer-Policy: same-origin
header: Vary: Origin
header: X-Content-Type-Options: nosniff
header: X-Frame-Options: DENY
header: Content-Length: 35
header: Connection: Close
ERROR: [Nebula] wendover-how-to-design-a-theme-park-to-take-tons-of-your-money: This video is only available for registered users. Use --cookies, --cookies-from-browser, --username and --password, or --netrc (watchnebula) to provide account credentials
  File "D:\DOS\YT-DLP~1\resources\WPy32\python-3.10.2\lib\site-packages\yt_dlp\extractor\common.py", line 642, in extract
    ie_result = self._real_extract(url)
  File "D:\DOS\YT-DLP~1\resources\WPy32\python-3.10.2\lib\site-packages\yt_dlp\extractor\nebula.py", line 246, in _real_extract
    video = self._fetch_video_metadata(slug)
  File "D:\DOS\YT-DLP~1\resources\WPy32\python-3.10.2\lib\site-packages\yt_dlp\extractor\nebula.py", line 239, in _fetch_video_metadata
    return self._call_nebula_api(f'https://content.watchnebula.com/video/{slug}/',
  File "D:\DOS\YT-DLP~1\resources\WPy32\python-3.10.2\lib\site-packages\yt_dlp\extractor\nebula.py", line 85, in _call_nebula_api
    self._perform_login()
  File "D:\DOS\YT-DLP~1\resources\WPy32\python-3.10.2\lib\site-packages\yt_dlp\extractor\nebula.py", line 148, in _perform_login
    self._nebula_api_token = self._retrieve_nebula_api_token(username, password)
  File "D:\DOS\YT-DLP~1\resources\WPy32\python-3.10.2\lib\site-packages\yt_dlp\extractor\nebula.py", line 67, in _retrieve_nebula_api_token
    return self._perform_nebula_auth(username, password)
  File "D:\DOS\YT-DLP~1\resources\WPy32\python-3.10.2\lib\site-packages\yt_dlp\extractor\nebula.py", line 23, in _perform_nebula_auth
    self.raise_login_required()
  File "D:\DOS\YT-DLP~1\resources\WPy32\python-3.10.2\lib\site-packages\yt_dlp\extractor\common.py", line 1140, in raise_login_required
    raise ExtractorError(msg, expected=True)


D:\DOS\yt-dlp_kit>set "EXIT_CODE=1" 

```

</details>

**Comment by stinerman** [comment]
I verified that I can download a few videos with this branch. The default filename is a bit verbose with the video id being part of it.

I can do some more thorough testing this weekend.

**Comment by hheimbuerger** [comment]
So weird. I looked at it again and now there's a `zype_id` field on the API. I could have sworn that hadn't been there last week. But of course the more likely explanation is that I simply missed it (while explicitly searching for it...).

**Comment by hheimbuerger** [comment]
I'm wondering if yt-dlp has an interface for delaying the generation of the list of formats and subtitles, only performing it on-demand.

That is, right now, if someone requests a playlist, for each entry in the list, the extractor will perform 1-2 HTTP requests. I believe that happens even if the ID is already in the archive, or if a video is filtered out by other criteria:

For example:
```
$ yt-dlp "https://nebula.app/adam-neely" --match-filter "title ~= (?i)music" --simulate --format=best
```

That channel has 112 videos, only a handful of which have the word "music" in the title. All the titles are returned by the three playlist page requests. But my current implementation also fetches formats and subtitles for all 112 videos.

**Comment by pukkandan** [comment]
There is a way, but I will need to see how feasible it is to do for this IE. Just remind me about this when u make the PR

**Comment by hheimbuerger** [comment]
> There is a way, but I will need to see how feasible it is to do for this IE. Just remind me about this when u make the PR

Do you want to point me to another extractor with a good ("reference") implementation that i could learn from?

**Comment by pukkandan** [comment]
Most playlist extractors (like the old Nebula extractor) return a URL first and then use it to extract more details of the video. For such extractors, this issue don't apply. The problem happens here because all extraction is done in a single step. So to address it, we could make the playlist IE return a `url_result`, either constructing one with the slug, or (if that is not possible) using a `ie_key:id` pattern.

I don't have an example for this off the top of my head, and I'm not very good at explaining code tbh. But I can look at the PR and write (pseudo?) code for how to do this, which you can then test/implement

**Comment by Lamieur** [comment]
While you ponder this, can we just get a fix for watching a single video? I'm all for fixing needless queries for channels/playlists, but that's not a show-stopper for fixing a "nothing works" condition ;)

(I know, I know - if you don't fix this now, it will stay like that forever. We've all been there :/)

**Comment by hheimbuerger** [comment]
@Lamieur The delay wasn't caused by the pondering, but rather because this breakage 'unfortunately' happened to coincide with a previously planned vacation on my end. 😉

I'm back since today and once the kids return to kindergarden on Tuesday, I'm hoping to be able to wrap this up quickly!

**Comment by Lamieur** [comment]
I just wanted to mention that Nebula works again with yt-dlp 2022.05.18 (unpatched), they must have rolled back that change?

**Comment by pukkandan** [comment]
The PR can still be merged (though with low priority), but closing the issue since it fixed itself

**Comment by hheimbuerger** [comment]
@pukkandan Makes sense! I also updated title and description of the PR to reflect that it's now a low prio refactoring, rather than an urgent fix.
